File: naive.rs
use std::collections::HashMap;
use std::f64::INFINITY;
use std::fs::File;
use std::io::{self, BufRead};

fn read_file(filename: &str) -> io::Result<io::Lines<io::BufReader<File>>> {
    let file = File::open(filename)?;
    return Ok(io::BufReader::new(file).lines());
}

#[derive(Debug)]
struct Measurement {
    min: f64,
    max: f64,
    sum: f64,
    cnt: f64,
}

impl Measurement {
    fn new() -> Measurement {
        Measurement {
            min: INFINITY,
            max: -INFINITY,
            sum: 0.0,
            cnt: 0.0,
        }
    }

    fn avg(&self) -> f64 {
        self.sum / self.cnt
    }

    fn format(&self) -> String {
        return format!("{:.1}/{:.1}/{:.1}", self.min, self.avg(), self.max);
    }
}

fn as_float(value_as_str: &str) -> f64 {
    return value_as_str.parse::<f64>().unwrap_or(0.0);
}

fn print_results(readings: HashMap<String, Measurement>) {
    let mut cities: Vec<_> = readings.iter().collect();

    cities.sort_by_key(|k| k.0);

    print!("{{");
    for (city, measurement) in cities.iter() {
        print!("{}={}, ", city, measurement.format());
    }

    print!("\x08\x08}} \n")
}

fn main() {
    match read_file("measurements.txt") {
        Err(_) => panic!("Unknown file."),
        Ok(file) => {
            let mut readings: HashMap<String, Measurement> = HashMap::new();

            for line in file.flatten() {
                let (location_as_str, value_as_str) = line.split_at(line.find(';').unwrap() + 1);

                let mut location = location_as_str.to_string();
                location.remove(location.len() - 1); // drop ;
                let value_as_float = as_float(value_as_str);

                // let reading: &&Measurement = readings.get(location).unwrap_or_else(|| &new_measurement());
                let reading = readings.entry(location).or_insert_with(&Measurement::new);

                if value_as_float > reading.max {
                    // reading.max.max(other)
                    reading.max = value_as_float
                }

                if value_as_float < reading.min {
                    reading.min = value_as_float
                }

                reading.sum += value_as_float;
                reading.cnt += 1.0;
            }

            print_results(readings);
        }
    }
}

File: naive.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Measurement struct {
	min float64
	max float64
	sum float64
	cnt float64
}

func NewMeasurement() Measurement {
	return Measurement{
		min: math.Inf(1),
		max: math.Inf(-1),
		sum: 0.0,
		cnt: 0.0,
	}
}

func (m Measurement) Avg() float64 {
	return m.sum / m.cnt
}

func (m Measurement) Format() string {
	return fmt.Sprintf("%.1f/%.1f/%.1f", m.min, m.Avg(), m.max)
}

func asFloat(valueAsString string) float64 {
	valueAsFloat, err := strconv.ParseFloat(valueAsString, 64)
	if err != nil {
		return 0.0
	}
	return valueAsFloat
}

func printResults(readings map[string]Measurement) {
	cities := make([]string, 0, len(readings))
	for city := range readings {
		cities = append(cities, city)
	}

	slices.Sort(cities)

	fmt.Print("{")
	for _, city := range cities {
		fmt.Printf("%s=%s, ", city, readings[city].Format())
	}
	fmt.Print("\b\b} \n")
}

func main() {

	file, err := os.Open("measurements.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)

	readings := make(map[string]Measurement)

	for scanner.Scan() {
		line := scanner.Text()

		location, valueAsString := split(line, ";")
		valueAsFloat := asFloat(valueAsString)

		reading, exists := readings[location]
		if !exists {
			reading = NewMeasurement()
		}

		if valueAsFloat > reading.max {
			reading.max = valueAsFloat
		}

		if valueAsFloat < reading.min {
			reading.min = valueAsFloat
		}

		reading.sum += valueAsFloat
		reading.cnt++

		readings[location] = reading
	}

	printResults(readings)
}

func split(value, sep string) (string, string) {
	parts := strings.Split(value, sep)
	return parts[0], parts[1]
}
